npeps/npep-133-fqdn-egress-selector.md

```diff
@@ -0,0 +1,442 @@
+hampers the readability of the network policies.
+
+## User Stories
+
+* As a cluster admin, I want to allow all Pods in the cluster to send traffic to
+  an external service specified by a well-known domain name. For example, all
+  Pods must be able to talk to `my-service.com`.
+
+* As a cluster admin, I want to allow Pods in the "monitoring" namespace to be
+  able to send traffic to a logs-sink, hosted at `logs-storage.com`
+
+* As a cluster admin, I want to allow all Pods in the cluster to send traffic to
+  any of the managed services provided by my Cloud Provider. Since the cloud
+  provider has a well known parent domain, I want to allow Pods to send traffic
+  to all sub-domains using a wild-card selector -- `*.my-cloud-provider.com`
+
+### Future User Stories
+
+These are some user stories we want to keep in mind, but due to limitations of
+the existing Network Policy API, cannot be implemented currently. The design
+goal in this case is to ensure we do not make these unimplementable down the
+line.
+
+* As a cluster admin, I want to block all cluster egress traffic by default, and
+  require namespace admins to create NetworkPolicies explicitly allowing egress
+  to the domains they need to talk to.
+
+  The Cluster admin would use a `BaselineAdminNetworkPolicy` object to switch
+  the default disposition of the cluster. Namespace admins would then use a FQDN
+  selector in the Kubernetes `NetworkPolicy` objects to allow `my-service.com`.
+  
+## API
+
+This NPEP proposes adding a new type of `AdminNetworkPolicyEgressPeer` called
+`FQDNPeerSelector` which allows specifying domain names.
+
+```golang
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+    // FQDN provides a way to specify domain names as peers.
+    //
+    // FQDN is only supported for ALLOW rules. In order to control access, FQDN
+    // ALLOW rules should be used with a lower priority egress deny -- this
+    // allows the admin to maintain an explicit "allowlist" of reachable
+   // domains.
+    //
+    // Support: Extended
+    //
+    // +optional
+    FQDN *FQDNPeerSelector `json:"fqdn,omitempty"`
+}
+
+type FQDNPeerSelector struct {
+    // Name specifies the exact Fully-Qualified Domain Name to select (the
+    // trailing '.' is optional)
+    //
+    // This field acts as an exact match. It doesn't match any other
+    // domains of that level or even any subdomains.
+    // Examples:
+    //   - `kubernetes.io` matches only `kubernetes.io`.
+    //     It does not match "www.kubernetes.io", "blog.kubernetes.io",
+    //     "my-kubernetes.io", or "wikipedia.org".
+    //   - `blog.kubernetes.io` matches only "blog.kubernetes.io".
+    //     It does not match "www.kubernetes.io" or "kubernetes.io".
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +kubebuilder:validation:Pattern=`^(([a-zA-z0-9]|[a-zA-Z0-9][-a-zA-Z0-9_]*[a-zA-Z0-9])\.)*([a-zA-z0-9]|[a-zA-Z0-9][-a-zA-Z0-9_]*[a-zA-Z0-9])*\.?$`
```

Comment:
I have a suggestions around regex:
`([a-zA-z0-9]|[a-zA-Z0-9][-a-zA-Z0-9_]*[a-zA-Z0-9])` can be replaced with 
`[a-zA-z0-9]([-a-zA-Z0-9_]*[a-zA-Z0-9])?`, do you think it is easier to read or does it miss some valid matches?
I am not sure I understand the last `*` in the expression, I think it can be just removed, otherwise `.` also matches.
This expression matches one-label names, like `com`, do we want to enforce at least 2-label match?

Reply:
Updated the regex. Thanks for the suggestions!